### ISSUE TITLE
Resource exhausted error for a pretty small network.

### ISSUE THREAD

**Issue by JosuaJoker** [keras-team/keras#14313]
Background: I constructed a multi-view cnn using one pretrained RestNet50 shared as a base model for feeding 20 image  of size 512 x 512 x 3 for one object in the net, in order to get after merging a feature representation of this 3D model and performing further analytic operations like classification or regression tasks. But I always get the **Resource exhausted error**. So I constructed a pretty SmallTestNet with only 9,537 parameters overall for testing, but getting the same issue with a batch of only (2x20x512x512x3), which are some MB's. 

 At one point a lot of memory is wasted. My source code see below. 

pciBusID: 0000:01:00.0 name: Quadro M1000M computeCapability: 5.0
coreClock: 1.0715GHz coreCount: 4 deviceMemorySize: 2.00GiB deviceMemoryBandwidth: 74.65GiB/s
2020-12-17 10:03:52.822759: I tensorflow/core/common_runtime/bfc_allocator.cc:950] InUse at 84e542800 of size 65545216 next 89
2020-12-17 10:03:52.828065: I tensorflow/core/common_runtime/bfc_allocator.cc:950] InUse at 8523c4c00 of size 82778624 next 18446744073709551615
2020-12-17 10:03:52.863198: I tensorflow/core/common_runtime/bfc_allocator.cc:955]      Summary of in-use Chunks by size: 
2020-12-17 10:03:52.900896: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 48 Chunks of size 256 totalling 12.0KiB
2020-12-17 10:03:52.911032: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 1 Chunks of size 1280 totalling 1.3KiB
2020-12-17 10:03:52.945792: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 2 Chunks of size 37632 totalling 73.5KiB
2020-12-17 10:03:52.988423: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 1 Chunks of size 72704 totalling 71.0KiB
2020-12-17 10:03:52.997040: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 19 Chunks of size 12582912 totalling 228.00MiB
2020-12-17 10:03:53.062703: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 1 Chunks of size 12583168 totalling 12.00MiB
2020-12-17 10:03:53.098472: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 15 Chunks of size 65545216 totalling 937.63MiB
2020-12-17 10:03:53.109608: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 1 Chunks of size 82778624 totalling 78.94MiB
2020-12-17 10:03:53.178818: I tensorflow/core/common_runtime/bfc_allocator.cc:958] 1 Chunks of size 107984896 totalling 102.98MiB
2020-12-17 10:03:53.220047: I tensorflow/core/common_runtime/bfc_allocator.cc:962] Sum Total of in-use chunks: 1.33GiB
2020-12-17 10:03:53.227935: I tensorflow/core/common_runtime/bfc_allocator.cc:964] total_region_allocated_bytes_: 1438344704 memory_limit_: 1438344806 available bytes: 102 curr_region_allocation_bytes_: 2876689920
2020-12-17 10:03:53.305283: I tensorflow/core/common_runtime/bfc_allocator.cc:970] Stats: 
Limit:                  1438344806
InUse:                  1425761792
MaxInUse:               1426183680
NumAllocs:                     209
MaxAllocSize:            251658240

2020-12-17 10:03:53.359601: W tensorflow/core/common_runtime/bfc_allocator.cc:429] *********************************xx****************************************************************x
2020-12-17 10:03:53.426945: W tensorflow/core/framework/op_kernel.cc:1655] OP_REQUIRES failed at conv_ops.cc:539 : Resource exhausted: OOM when allocating tensor with shape[4,64,253,253] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
`
BATCH_SIZE = 4
views=20
channels = 3
image_size = 512
IMAGE_SIZE = [image_size, image_size]
IMAGES_PER_MODEL = list(range(0,views))
AUTOTUNE = tf.data.experimental.AUTOTUNE
def SmallTestResNet():
    img_input = layers.Input(shape=(image_size,image_size,channels))
    x = layers.Conv2D(64, (7, 7),
                      strides=(2, 2),
                      padding='valid',
                      kernel_initializer='he_normal',
                      name='conv1')(img_input)
    x = layers.Activation('relu')(x)
    x = layers.GlobalAveragePooling2D()(x)
    model = Model(img_input, x, name='SmallTestResNet')
    return model

def create_model():
    nets, inputs=[], []

    # Set up base model
    # base_ResNet50 = ResNet50(weights='imagenet', include_top= False, input_shape=(512, 512, 3))
    #for layer in base_ResNet50.layers[:-4]:
    # for layer in base_ResNet50.layers:
    #     layer.trainable = False
    base_ResNet50 = SmallTestResNet()
    
    # Construct mvcnn
    inputs = Input(shape=(views,image_size,image_size,channels))
    for x in tf.split(inputs,num_or_size_splits=views, axis=1):
        x = tf.reshape(x,[-1,image_size,image_size,channels])
        y = base_ResNet50(x)
        nets.append(y)
    maxpooling = tf.reduce_max(nets, [0])
    flatten = Flatten()(maxpooling)
    dense_1 = Dense(10,activation='sigmoid')(flatten)
    predictions = Dense(1,activation='sigmoid')(dense_1)
    model = Model(inputs=inputs, outputs=predictions)
    model.compile(loss='binary_crossentropy', optimizer='adam',
            metrics=['accuracy'])
    return model

def train_model():
    create_tfrecords(tfrecord_path)
    model = create_model()
    # Get data from tfrecord    
    train_dataset = get_dataset(tfrecord_path)
    # Fit model with data
    model.fit(train_dataset,steps_per_epoch = 20, epochs=5)

def read_tfrecord(bin_data):
    label_seq={}
    for i in feature_map_dict:
        label_seq[i] = tf_input_feature_selector(feature_map_dict[i])
    img_seq = {'images': tf.io.FixedLenSequenceFeature([], dtype=tf.string)}
    cont, seq = tf.io.parse_single_sequence_example(serialized=bin_data,
                                                context_features=label_seq,
                                                sequence_features=img_seq)

    # Get the image as raw bytes.
    image_raw = seq['images']
    images = decode_image_raw(image_raw)    
    images = tf.reshape(images, [views,image_size,image_size,channels])
    images = preprocess_input(images)
    label = cont['label']
    return images, label

def get_dataset(tfrecord_path):
    dataset = tf.data.TFRecordDataset(filenames=tfrecord_path)
    dataset = dataset.map(read_tfrecord)
    #dataset = dataset.repeat(-1)
    #dataset = dataset.shuffle(2048)
    dataset = dataset.prefetch(buffer_size=AUTOTUNE)
    dataset = dataset.batch(BATCH_SIZE)
    return dataset`

**Comment by JosuaJoker** [comment]
Only with a batch_size of 1 it works. Does there exist some performance hints. Thank you very much in advance.

**Comment by jvishnuvardhan** [comment]
@JosuaJoker Sorry for the late response. Is this still an issue for you?

Can you please share a simple standalone code to reproduce the issue? Please format the code and share any data (public) with the standalone code. Thanks!

**Comment by JosuaJoker** [comment]
At the moment, I am not able, because it is long ago and I distroied the code very much to get it working. Less views and a much smaller resolution was unfortunatlly the key. I will be able to test my code soon on a much stronger machine with more views and higher resolution again and let you know if the issue still exists. But from my feeling, I had the impression that the performance with pure tensorflow is very much better than with keras and tensorflow as backend, so perhaps you have some good hints for me to improve the performance with keras and optimise the code, in order to give you a more solid statement.
